### PR TITLE
Use C SIMD implementation instead of libgmp

### LINKED ISSUE #63: SIMD implementation for `zipBits`/`invertBits`
`bitvec` has the option to use libgmp for faster implementations of `zipBits` and `invertBits`.

I noticed that you can get even better performance with a C implementation that uses SIMD (this is really easy through OpenMP, with `#pragma omp simd`), for example:

```c
void omp_and(uint8_t *dest, const uint8_t *src1, const uint8_t *src2, size_t len) {
    #pragma omp simd
    for (size_t i = 0; i < len; i++) {
        dest[i] = src1[i] & src2[i];
    }
}
```

OpenMP is supported by all standard C compilers. The SIMD versions scale better, being about twice as fast as the libgmp versions for higher amounts of bits in the benchmark (looking at its source code, libgmp doesn't seem to use SIMD optimized implementations for this, so this makes sense).

<details><summary>Benchmark results</summary>

I benched the SIMD version in the `invert`, `intersection` and `union` benchmarks (as those use the optimized implementations), with the baseline being the libgmp version.

```
All
  invert
    32
      Bit:    OK (0.25s)
        27.6 ns ± 2.0 ns,       same as baseline
      BitTS:  OK (0.23s)
        25.2 ns ± 1.5 ns, 0.91x, 21% less than baseline
      Vector: OK (0.29s)
        65.1 ns ± 5.0 ns, 2.36x,       same as baseline
    64
      Bit:    OK (0.24s)
        27.3 ns ± 1.6 ns,       same as baseline
      BitTS:  OK (0.23s)
        25.2 ns ± 1.4 ns, 0.92x, 19% less than baseline
      Vector: OK (0.25s)
        109  ns ± 6.3 ns, 4.01x,       same as baseline
    128
      Bit:    OK (0.26s)
        28.2 ns ± 1.4 ns, 10% less than baseline
      BitTS:  OK (0.24s)
        25.9 ns ± 1.9 ns, 0.92x,       same as baseline
      Vector: OK (0.23s)
        200  ns ±  11 ns, 7.09x,       same as baseline
    256
      Bit:    OK (0.26s)
        29.7 ns ± 1.9 ns,       same as baseline
      BitTS:  OK (0.25s)
        27.5 ns ± 1.7 ns, 0.93x,       same as baseline
      Vector: OK (0.22s)
        380  ns ±  22 ns, 12.77x,       same as baseline
    512
      Bit:    OK (0.26s)
        28.6 ns ± 1.3 ns, 13% less than baseline
      BitTS:  OK (0.24s)
        26.4 ns ± 1.6 ns, 0.92x, 17% less than baseline
      Vector: OK (0.21s)
        750  ns ±  43 ns, 26.20x,       same as baseline
    1024
      Bit:    OK (0.14s)
        28.8 ns ± 2.6 ns, 22% less than baseline
      BitTS:  OK (0.25s)
        26.8 ns ± 1.6 ns, 0.93x, 23% less than baseline
      Vector: OK (0.21s)
        1.48 μs ±  93 ns, 51.49x,       same as baseline
    2048
      Bit:    OK (0.29s)
        31.7 ns ± 1.5 ns, 31% less than baseline
      BitTS:  OK (0.14s)
        30.9 ns ± 2.8 ns, 0.98x, 23% less than baseline
      Vector: OK (0.40s)
        2.96 μs ± 126 ns, 93.48x,       same as baseline
    4096
      Bit:    OK (0.19s)
        38.7 ns ± 2.8 ns, 36% less than baseline
      BitTS:  OK (0.18s)
        38.5 ns ± 3.3 ns, 0.99x, 36% less than baseline
      Vector: OK (0.87s)
        12.6 μs ± 389 ns, 325.57x,       same as baseline
    8192
      Bit:    OK (0.44s)
        48.9 ns ± 4.0 ns, 49% less than baseline
      BitTS:  OK (0.43s)
        47.3 ns ± 1.6 ns, 0.97x, 51% less than baseline
      Vector: OK (0.41s)
        46.7 μs ± 3.3 μs, 955.37x,       same as baseline
    16384
      Bit:    OK (0.18s)
        71.4 ns ± 6.0 ns, 61% less than baseline
      BitTS:  OK (0.20s)
        79.0 ns ± 6.6 ns, 1.11x, 57% less than baseline
      Vector: OK (0.25s)
        114  μs ± 7.6 μs, 1595.65x,       same as baseline
  intersection
    32
      Bit:    OK (0.22s)
        92.3 ns ± 5.2 ns,       same as baseline
      BitTS:  OK (0.23s)
        105  ns ± 5.5 ns, 1.14x, 15% more than baseline
      Vector: OK (0.15s)
        251  ns ±  23 ns, 2.72x,       same as baseline
      IntSet: OK (0.16s)
        16.5 ns ± 1.6 ns, 0.18x,       same as baseline
    64
      Bit:    OK (0.22s)
        92.4 ns ± 5.8 ns,       same as baseline
      BitTS:  OK (0.23s)
        100  ns ± 5.4 ns, 1.08x, 10% more than baseline
      Vector: OK (0.26s)
        453  ns ±  38 ns, 4.91x,       same as baseline
      IntSet: OK (0.16s)
        16.6 ns ± 1.5 ns, 0.18x,       same as baseline
    128
      Bit:    OK (0.22s)
        95.9 ns ± 9.6 ns,       same as baseline
      BitTS:  OK (0.22s)
        92.9 ns ± 7.5 ns, 0.97x,       same as baseline
      Vector: OK (0.25s)
        871  ns ±  58 ns, 9.08x,       same as baseline
      IntSet: OK (0.65s)
        37.5 ns ± 3.1 ns, 0.39x,       same as baseline
    256
      Bit:    OK (0.22s)
        95.8 ns ± 7.1 ns,       same as baseline
      BitTS:  OK (0.22s)
        94.5 ns ± 5.1 ns, 0.99x,       same as baseline
      Vector: OK (0.24s)
        1.70 μs ±  97 ns, 17.72x,       same as baseline
      IntSet: OK (0.18s)
        77.2 ns ± 5.4 ns, 0.81x,       same as baseline
    512
      Bit:    OK (0.22s)
        91.9 ns ± 6.1 ns,       same as baseline
      BitTS:  OK (0.21s)
        91.6 ns ± 5.5 ns, 1.00x,       same as baseline
      Vector: OK (0.24s)
        3.41 μs ± 176 ns, 37.10x,       same as baseline
      IntSet: OK (0.21s)
        176  ns ±  12 ns, 1.92x,       same as baseline
    1024
      Bit:    OK (0.22s)
        92.1 ns ± 6.1 ns,       same as baseline
      BitTS:  OK (0.21s)
        91.7 ns ± 5.5 ns, 1.00x,  9% less than baseline
      Vector: OK (0.27s)
        7.67 μs ± 351 ns, 83.24x,       same as baseline
      IntSet: OK (0.77s)
        351  ns ± 8.9 ns, 3.81x,       same as baseline
    2048
      Bit:    OK (0.42s)
        92.1 ns ± 3.4 ns, 16% less than baseline
      BitTS:  OK (0.23s)
        94.9 ns ± 7.4 ns, 1.03x, 16% less than baseline
      Vector: OK (0.20s)
        22.8 μs ± 1.3 μs, 247.50x,       same as baseline
      IntSet: OK (0.23s)
        773  ns ±  45 ns, 8.39x,       same as baseline
    4096
      Bit:    OK (0.24s)
        98.7 ns ± 6.4 ns, 26% less than baseline
      BitTS:  OK (0.24s)
        98.4 ns ± 5.7 ns, 1.00x, 28% less than baseline
      Vector: OK (0.14s)
        60.6 μs ± 6.0 μs, 613.72x,       same as baseline
      IntSet: OK (0.23s)
        1.57 μs ±  90 ns, 15.93x,       same as baseline
    8192
      Bit:    OK (0.27s)
        112  ns ± 5.4 ns, 42% less than baseline
      BitTS:  OK (0.27s)
        112  ns ± 5.6 ns, 1.01x, 42% less than baseline
      Vector: OK (0.15s)
        126  μs ±  12 μs, 1130.02x,       same as baseline
      IntSet: OK (0.25s)
        3.26 μs ± 188 ns, 29.19x,       same as baseline
    16384
      Bit:    OK (0.18s)
        126  ns ±  12 ns, 58% less than baseline
      BitTS:  OK (0.18s)
        128  ns ±  11 ns, 1.02x, 57% less than baseline
      Vector: OK (0.29s)
        259  μs ±  13 μs, 2064.13x,       same as baseline
      IntSet: OK (1.77s)
        6.55 μs ± 118 ns, 52.10x,       same as baseline
  union
    32
      Bit:    OK (0.23s)
        89.6 ns ± 8.3 ns,       same as baseline
      BitTS:  OK (0.23s)
        90.2 ns ± 6.0 ns, 1.01x,       same as baseline
      Vector: OK (0.27s)
        226  ns ±  10 ns, 2.53x,       same as baseline
      IntSet: OK (0.18s)
        15.9 ns ± 1.5 ns, 0.18x,       same as baseline
    64
      Bit:    OK (0.22s)
        89.4 ns ± 5.2 ns,       same as baseline
      BitTS:  OK (0.23s)
        89.8 ns ± 5.4 ns, 1.01x,       same as baseline
      Vector: OK (0.26s)
        429  ns ±  22 ns, 4.81x,       same as baseline
      IntSet: OK (0.18s)
        16.0 ns ± 1.4 ns, 0.18x,       same as baseline
    128
      Bit:    OK (0.23s)
        89.9 ns ± 5.8 ns,       same as baseline
      BitTS:  OK (0.23s)
        90.4 ns ± 5.7 ns, 1.01x,       same as baseline
      Vector: OK (0.25s)
        836  ns ±  43 ns, 9.29x,       same as baseline
      IntSet: OK (0.18s)
        35.9 ns ± 2.7 ns, 0.40x,       same as baseline
    256
      Bit:    OK (0.23s)
        92.6 ns ± 6.3 ns,       same as baseline
      BitTS:  OK (0.23s)
        91.8 ns ± 6.0 ns, 0.99x,       same as baseline
      Vector: OK (0.24s)
        1.63 μs ±  84 ns, 17.64x,       same as baseline
      IntSet: OK (0.20s)
        74.9 ns ± 5.4 ns, 0.81x,       same as baseline
    512
      Bit:    OK (0.23s)
        90.4 ns ± 6.1 ns,       same as baseline
      BitTS:  OK (0.23s)
        90.6 ns ± 5.5 ns, 1.00x,       same as baseline
      Vector: OK (0.25s)
        3.27 μs ± 267 ns, 36.17x,       same as baseline
      IntSet: OK (0.40s)
        172  ns ± 7.2 ns, 1.91x,       same as baseline
    1024
      Bit:    OK (0.23s)
        89.3 ns ± 5.5 ns,       same as baseline
      BitTS:  OK (0.23s)
        90.9 ns ± 5.3 ns, 1.02x,  8% less than baseline
      Vector: OK (0.25s)
        6.79 μs ± 376 ns, 76.02x,       same as baseline
      IntSet: OK (0.41s)
        355  ns ±  26 ns, 3.98x,       same as baseline
    2048
      Bit:    OK (0.23s)
        92.2 ns ± 5.5 ns, 13% less than baseline
      BitTS:  OK (0.22s)
        93.3 ns ± 5.4 ns, 1.01x, 14% less than baseline
      Vector: OK (0.20s)
        20.9 μs ± 1.4 μs, 226.96x,       same as baseline
      IntSet: OK (0.24s)
        771  ns ±  46 ns, 8.37x,       same as baseline
    4096
      Bit:    OK (0.25s)
        96.6 ns ± 7.0 ns, 28% less than baseline
      BitTS:  OK (0.25s)
        97.1 ns ± 6.0 ns, 1.01x, 28% less than baseline
      Vector: OK (0.26s)
        56.3 μs ± 3.8 μs, 582.50x,       same as baseline
      IntSet: OK (0.25s)
        1.58 μs ± 108 ns, 16.36x,       same as baseline
    8192
      Bit:    OK (0.16s)
        105  ns ±  10 ns, 45% less than baseline
      BitTS:  OK (0.26s)
        104  ns ± 5.5 ns, 0.99x, 46% less than baseline
      Vector: OK (0.28s)
        122  μs ±  10 μs, 1160.93x,       same as baseline
      IntSet: OK (0.25s)
        3.23 μs ± 190 ns, 30.66x,       same as baseline
    16384
      Bit:    OK (0.18s)
        114  ns ±  10 ns, 61% less than baseline
      BitTS:  OK (0.18s)
        114  ns ±  11 ns, 1.00x, 62% less than baseline
      Vector: OK (0.16s)
        260  μs ±  24 μs, 2285.21x,       same as baseline
      IntSet: OK (0.26s)
        6.55 μs ± 417 ns, 57.52x,       same as baseline
```

</details>

My implementation can be found at https://github.com/konsumlamm/bitvec/tree/simd. It could also be expanded to be applicable in cases where the offset is non-zero. A similar implementation should also work for `zipInPlace`/`invertInPlace` and perhaps also others.

Would you be interested in a PR for this?

### PR DESCRIPTION
**Pull request by konsumlamm**
Resolves #63.

This is basically a drop-in replacement for `libgmp`. The only thing currently missing is an equivalent of `mpnPopcount`.

As you can see, using OpenMP for SIMD is pretty easy and doesn't require any intrinsics, keeping the code simple and easy to understand. OpenMP only produces basic SIMD instruction that should be available on every modern CPU (see https://godbolt.org/z/zqz3nEaWj for the generated assembly).

### Future Work

* make operations work for bit vectors with offset
* support more immutable operations (e.g. `bitIndex`, `nthBitIndex`, `countBits`, `selectBits`, `excludeBits`)
* support mutable operations (`...InPlace` variants)

### CODE REVIEW

File: cbits/bitvec_simd.c [hunk 1]
```diff
@@ -0,0 +1,65 @@
+#include <inttypes.h>
+#include <stddef.h>
+
+void omp_com(uint8_t *dest, uint8_t *src, size_t len) {
```
Comment (Bodigrim) [inline cbits/bitvec_simd.c:4]: I suspect that we need
```suggestion
void _hs_bitvec_omp_com(uint8_t *dest, uint8_t *src, size_t len) {
```
and similar below.

File: src/Data/Bit/Immutable.hs [hunk 2]
```diff
@@ -680,10 +670,6 @@ unsafeNthTrueInWord l w = countTrailingZeros (pdep (1 `shiftL` (l - 1)) w)
 -- @since 0.1
 countBits :: U.Vector Bit -> Int
 countBits (BitVec _ 0 _)                      = 0
-#if UseLibGmp
-countBits (BitVec 0 len arr) | modWordSize len == 0 =
-  fromIntegral (mpnPopcount arr (divWordSize len))
```
Comment (Bodigrim) [inline src/Data/Bit/Immutable.hs:685]: Could you please implement `countBits` as well?

File: bitvec.cabal [hunk 3]
```diff
@@ -94,11 +93,12 @@ library
   else
     build-depends: ghc-bignum
 
-  if flag(libgmp)
-    extra-libraries: gmp
-    cpp-options: -DUseLibGmp
+  if flag(simd)
+    c-sources: cbits/bitvec_simd.c
+    cc-options: -O2 -fopenmp
```
Comment (Bodigrim) [inline bitvec.cabal:98]: Let's make `cabal check` happy.
```suggestion
    cc-options: -fopenmp
```

File: cbits/bitvec_simd.c (hunk 1 above)
Comment (konsumlamm) [reply cbits/bitvec_simd.c:4]: You mean to avoid naming conflicts?

Comment (Bodigrim) [reply cbits/bitvec_simd.c:4]: Yes. At least this is what `bytestring` and `text` do.

File: bitvec.cabal [hunk 4]
```diff
@@ -55,11 +55,10 @@ source-repository head
...
 
-flag libgmp
+flag simd
   description:
-    Link against the GMP library for the ultimate performance of
-    `zipBits`, `invertBits` and `countBits`. Users are strongly encouraged
-    to enable this flag whenever possible.
+    Use a C SIMD implementation for the ultimate performance of `zipBits` and `invertBits`.
+    Users are strongly encouraged to enable this flag whenever possible.
   default: False
```
Comment (Bodigrim) [inline bitvec.cabal:62]: Shall we enable it by default?

File: src/Data/Bit/Immutable.hs (hunk 2 above)
Comment (konsumlamm) [reply src/Data/Bit/Immutable.hs:685]: I managed to come up with a version that has about the same speed as the GMP version.

File: bitvec.cabal (hunk 4 above)
Comment (konsumlamm) [reply bitvec.cabal:62]: Hmm, wouldn't that break with the JS or WASM backend (or in the rare case that the default C compiler doesn't support OpenMP)?

Comment (Bodigrim) [reply bitvec.cabal:62]: I would not worry about them, they are too boutique. Until there is a CI job to witness the contrary, we can effectively assume that they are broken anyways.

Flipping a package flag is likely the least of obstacles for GHC-to-JS compilation right now. We can potentially say `if flag(openmp) && !arch(javascript) ...`, if Cabal 3.10 already supports it.

**Comment by konsumlamm** [comment]
IIUC, the byte arrays are guaranteed to always have a multiple of 32 (or 64) bits? The current `_hs_bitvec_popcount` implementation relies on that the input array can be interpreted as an array of 32 bit words.

**Comment by Bodigrim** [comment]
> IIUC, the byte arrays are guaranteed to always have a multiple of 32 (or 64) bits?

Yes, 32 or 64 bits depending on your platform word size.

**Comment by konsumlamm** [comment]
> LGTM, thanks. Feel free to squash / reorder if you wish before I merge.

You can just *Squash and Merge*, I'm fine with the history being available in this PR.

**Comment by Bodigrim** [comment]
It's fairly clean history, I'm happy to retain it as is. 

Thanks, that's a great contribution!

**Pull request merged**